// angular6-material-table/esm2015/lib/angular6-material-table-element.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import cloneDeep from 'lodash.clonedeep';
/**
 * @template T
 */
export class TableElement {
    /**
     * @param {?} init
     */
    constructor(init) {
        Object.assign(this, init);
        this.fillValidatorFromData();
        if (this.validator) {
            if (this.editing) {
                this.validator.enable();
            }
            else {
                this.validator.disable();
            }
        }
    }
    /**
     * @return {?}
     */
    get currentData() {
        //express hack
        if (this.validator) {
            return Object.assign(this._currentData, this.validator.getRawValue());
        }
        else {
            return this._currentData;
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set currentData(value) {
        this._currentData = value;
        this.fillValidatorFromData();
    }
    /**
     * @return {?}
     */
    fillValidatorFromData() {
        if (this.validator) {
            let /** @type {?} */ formData = {};
            for (let /** @type {?} */ key in this.validator.controls) {
                formData[key] = this._currentData[key];
            }
            this.validator.setValue(formData);
        }
    }
    /**
     * @return {?}
     */
    delete() {
        this.source.delete(this.id);
    }
    /**
     * @return {?}
     */
    confirmEditCreate() {
        this.originalData = undefined;
        if (this.id == -1)
            return this.source.confirmCreate(this);
        else
            return this.source.confirmEdit(this);
    }
    /**
     * @return {?}
     */
    startEdit() {
        this.originalData = cloneDeep(this.currentData);
        this.editing = true;
        this.validator.enable();
    }
    /**
     * @return {?}
     */
    cancelOrDelete() {
        if (this.id == -1 || !this.editing)
            this.delete();
        else {
            this.currentData = this.originalData;
            this.editing = false;
            this.validator.disable();
        }
    }
}
function TableElement_tsickle_Closure_declarations() {
    /** @type {?} */
    TableElement.prototype.id;
    /** @type {?} */
    TableElement.prototype.editing;
    /** @type {?} */
    TableElement.prototype._currentData;
    /** @type {?} */
    TableElement.prototype.originalData;
    /** @type {?} */
    TableElement.prototype.source;
    /** @type {?} */
    TableElement.prototype.validator;
}
