// angular6-material-table/esm5/lib/angular6-material-table-element.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import cloneDeep from 'lodash.clonedeep';
/**
 * @template T
 */
var /**
 * @template T
 */
TableElement = /** @class */ (function () {
    function TableElement(init) {
        Object.assign(this, init);
        this.fillValidatorFromData();
        if (this.validator) {
            if (this.editing) {
                this.validator.enable();
            }
            else {
                this.validator.disable();
            }
        }
    }
    Object.defineProperty(TableElement.prototype, "currentData", {
        get: /**
         * @return {?}
         */
        function () {
            //express hack
            if (this.validator) {
                return Object.assign(this._currentData, this.validator.getRawValue());
            }
            else {
                return this._currentData;
            }
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._currentData = value;
            this.fillValidatorFromData();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TableElement.prototype.fillValidatorFromData = /**
     * @return {?}
     */
    function () {
        if (this.validator) {
            var /** @type {?} */ formData = {};
            for (var /** @type {?} */ key in this.validator.controls) {
                formData[key] = this._currentData[key];
            }
            this.validator.setValue(formData);
        }
    };
    /**
     * @return {?}
     */
    TableElement.prototype.delete = /**
     * @return {?}
     */
    function () {
        this.source.delete(this.id);
    };
    /**
     * @return {?}
     */
    TableElement.prototype.confirmEditCreate = /**
     * @return {?}
     */
    function () {
        this.originalData = undefined;
        if (this.id == -1)
            return this.source.confirmCreate(this);
        else
            return this.source.confirmEdit(this);
    };
    /**
     * @return {?}
     */
    TableElement.prototype.startEdit = /**
     * @return {?}
     */
    function () {
        this.originalData = cloneDeep(this.currentData);
        this.editing = true;
        this.validator.enable();
    };
    /**
     * @return {?}
     */
    TableElement.prototype.cancelOrDelete = /**
     * @return {?}
     */
    function () {
        if (this.id == -1 || !this.editing)
            this.delete();
        else {
            this.currentData = this.originalData;
            this.editing = false;
            this.validator.disable();
        }
    };
    return TableElement;
}());
/**
 * @template T
 */
export { TableElement };
function TableElement_tsickle_Closure_declarations() {
    /** @type {?} */
    TableElement.prototype.id;
    /** @type {?} */
    TableElement.prototype.editing;
    /** @type {?} */
    TableElement.prototype._currentData;
    /** @type {?} */
    TableElement.prototype.originalData;
    /** @type {?} */
    TableElement.prototype.source;
    /** @type {?} */
    TableElement.prototype.validator;
}
